[1단계 - 자동차 경주 구현] 져니(이지원) 미션 제출합니다.

Cl8D opened the pull request:
안녕하세요! 우테코 5기 백엔드 크루 져니입니다 ⛄️
준팍과 함께 첫 페어 프로그래밍 너무 재밌게 진행했습니다!
코드에서 부족한 점, 수정했으면 좋겠는 점 아낌없이 피드백 해주시면 감사드리겠습니다!
잘 부탁드립니다. 감사합니다!

Cl8D committed:
test: string에 대한 테스트 코드 작성

Cl8D committed:
test: set에 대한 테스트 코드 작성

Cl8D committed:
docs: 기능 명세서 및 도메인 설계 작성

Cl8D committed:
feat: 자동차 전진 및 결과 출력 기능 추가

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
feat: 자동차 이름에 대한 검증 기능 및 예외 메시지 상수 추가

- 자동차 이름의 길이에 대한 검증 (5글자 초과)
- 자동차 이름의 형식에 대한 검증 (공백인지 체크)

Cl8D committed:
refactor: EnumUtil 인터페이스로 리팩토링

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
fix: 자동차 이동 시 랜덤 값 판단 로직을 제거

Cl8D committed:
feat: 자동차 경주에 대한 컨트롤러 추가

Cl8D committed:
feat: 예외 발생 시 재입력 및 자동차 이름 입력 기능 추가

Cl8D committed:
feat: 자동차 리스트를 관리하는 일급 컬렉션 추가

Cl8D committed:
refactor: EnumUtil의 반환 클래스 추가

Cl8D committed:
refactor: Car의 생성자에 대해 private로 변경 및 static factory method로 변경

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
refactor: 메서드 위치 변경

Cl8D committed:
feat: 자동차 경주의 시도할 횟수 입력 및 예외 검증 기능 추가

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
test: Cars 도메인에 대한 테스트 추가

- 정상적인 차량 이름이 들어오면 예외 발생 X
- 차량 이름에 null이 들어오면 예외 발생

Cl8D committed:
test: Car 도메인에 대한 테스트 추가

- 자동차 이름이 정상적으로 들어왔을 경우 예외가 발생하지 않는다.
- 자동차 이름이 5글자 초과이거나 공백일 경우 예외가 발생한다.

Cl8D committed:
test: Race 도메인에 대한 테스트 추가

- 정상적인 시도 횟수가 들어왔을 경우 예외가 발생하지 않는다.
- 시도 횟수가 정수값이 아니거나, 0 이하일 경우 예외가 발생한다.

Cl8D committed:
refactor: 도메인에 대한 상수로 클래스명 변경 및 시도 횟수에 대한 enum 추가

Cl8D committed:
feat: 0~9 사이의 숫자를 랜덤하게 생성하는 제너레이터 추가

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
feat: 자동차의 초기 주행 상태 추가 및 메시지 출력 시 static method 사용

Cl8D committed:
feat: 자동차 경주 로직 추가

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
feat: 랜덤 숫자를 매개변수로 받고, private method로 변경

Cl8D committed:
test: 랜덤 숫자에 따른 자동차 이동 여부에 대한 테스트 추가

Cl8D committed:
feat: 최종 우승 결과를 출력하는 기능 추가

- 주행 거리가 가장 많은 자동차의 이름 리스트를 출력하도록 추가

Cl8D committed:
docs: 구현 기능에 대해 체크

Cl8D committed:
feat: 차 이름에 대한 중복값 체크 로직 추가

Cl8D committed:
feat: ConsoleException 추가 및 예외 발생 시 재입력받을 수 있도록 로직 수정

Cl8D committed:
test: 우승한 자동차의 이름을 검증하는 테스트 추가

- 주행 거리에 따라 우승한 자동차를 뽑고, 해당 자동차가 맞는지 검증하는 테스트 추가

Cl8D committed:
style: 코드 포맷팅 변경

Cl8D committed:
docs: 기능 명세서 최종 수정

Cl8D committed:
docs: 체크박스 형식 수정

-깃허브에서 깨져보이는 문제 해결!

jnsorn commented on docs/README.md:
```diff
@@ -0,0 +1,98 @@
+
+---
+
+## ✔️ 기능 명세서
```
기능명세서를 꼼꼼하게 잘 작성해주셨네요!
커밋을 보니 기능명세서를 토대로 작업을 잘 진행해주신 것 같아요. 💯

jnsorn commented on docs/README.md:
```diff
@@ -0,0 +1,98 @@
+
+---
+
+## ✔️ 도메인 설계
```
어떤 객체가 어떤 책임을 갖고 있는지도 작성해주셨네요! 꼼꼼함..😮👍

jnsorn commented on src/main/java/racingcar/enumType/DomainConstant.java:
```diff
@@ -0,0 +1,26 @@
+
+    CAR_FORWARD_NUMBER(4)
+    , CAR_NAME_MAX_LENGTH(5)
+    , RACE_MIN_TRY_COUNT(0);
```
입출력 메세지 뿐만 아니라 도메인 제약사항까지 enum으로 한 곳에서 관리하고 있군요!
메세지와 제약사항들을 한 곳에서 관리하면서 어떤 점이 좋았고, 어떤 점이 불편했는지 듣고 싶어요. ☺️

Cl8D replied:
저는 기본적으로 상수를 enum으로 관리하는 편인데, enum으로 관리하면서 좋았던 점은 다음과 같이 말씀드릴 수 있을 것 같아요!

1. 입출력 메시지의 경우 사용자에게 보여지는 메시지인 만큼 **변화가 자주 일어날 수 있는 부분**이라고 생각이 들어요! 이때, 해당 메시지를 다양한 곳에서 참조하는 경우 모든 곳의 코드를 바꿔줘야 하는데 enum으로 관리해주면 **한 곳에서 관리**할 수 있어서 관리 차원에서 용이하다는 생각이 들어요.
2. 도메인 제약 사항의 경우, 이번 과제에서는 대부분 숫자값을 가지고 있는데 (흔히 ‘매직 넘버’라고 하죠!) 해당 숫자값에 대해서 그대로 코드에 사용하게 되면 코드를 읽는 입장에서 해당 넘버가 무슨 의미를 가지고 있는지 한 눈에 알아보기가 힘들 것이라고 생각했어요. 그래서 enum을 통해 해당 숫자에 대해서 **별도의 명칭으로 관리**하고, 또 마찬가지로 **변경이 용이**할 수 있도록 한 곳에서 관리하려고 사용하는 것 같아요!

다만, 현재는 도메인 제약사항 자체를 하나의 ‘DomainConstant’에서 관리하다 보니까, 추후 확장성을 고려해봤을 때 도메인의 크기가 커지게 된다면 내부에 들어가야 하는 값이 많아져서 복잡도가 증가할 것 같아요. 그래서 이번 리팩토링 때 제약사항들은 **각 도메인에 맞게 한 번 더 분리 작업을 진행**해보려고 합니다!

또한, 도메인 내에서 상수를 static final로 관리한다면 내부적으로 조금 더 코드를 작성하는 입장에서는 덜 귀찮을(?) 수도 있을 것 같고, 단순히 상수가 아니라 enum 역시 별도의 타입(이라고 표현해도 될지 모르겠지만)이기 때문에 메모리를 더 잡아먹을 것 같아요…!

jnsorn replied:
져니 의견 들려줘서 고마워요! 👍 
개인적으로 저는 도메인 내에서 상수로 관리하는 방식을 선호하는데요, 그 이유는 아래와 같아요. 

1. 예외 메세지를 예외 로직과 함께 두면 가독성이 올라간다.
2. 특정 예외상황에 대한 로직들을 한 곳에서 관리할 수 있다.
3. 도메인 규칙이 변경된 경우, 해당 도메인 규칙을 관리하는 곳에서만 코드가 변경되면 된다. (ex. 이름 조건이 변경되었다면 Name클래스만 변경)

특히, 하나의 메세지나 제약사항을 여러 클래스에서 사용한다는 것은 동일한 로직이 여기저기 흩어져 있다는 신호일 수도 있어요. 그런 경우에 해당 로직은 어느 객체의 역할일지 고민해보는게 좋겠죠! 

물론 저의 생각이니 참고만 해주시고, 져니가 선호하는 방식으로 구현해주세요. 😃

jnsorn commented on src/main/java/racingcar/util/InputUtil.java:
```diff
@@ -0,0 +1,39 @@
+        OutputView.printMessage(GET_CAR_NAMES_MESSAGE.getValue());
+        String carNames = inputView.getUserInput();
+        return Cars.of(carNames);
+    }
```
> **InputUtil**: 사용자에게 필요한 값을 입력받고, 예외 발생 시 재입력받도록 하는 유틸 클래스. 

져니가 작성한 문서에 따르면 InputUtil은 사용자로부터 값을 입력받는 클래스네요! 하지만 코드를 보면 Cars를 생성하는 책임까지 갖고 있는 것 같이 보여요. 🥲
InputUtil의 역할과 책임은 무엇일지 고민해보고 의견 들려주실 수 있나요~?

Cl8D replied:
InputUtil이라고 지었던 이유는 InputView에서는 사용자에게 단순히 콘솔을 통해서 입력을 받는 기능만 넣어두고, InputUtil는 **InputView의 일종의 진입점**으로서 다른 곳에서 참조하도록 진행하려고 별도의 유틸 클래스로 분리했어요…!  

+) 원래는 InputService라고 두어야 하나… 고민을 했는데, 이 [포스팅](https://blog.vhoyer.dev/posts/difference-between-services-and-utils/)을 보면서 service의 경우 ‘범위가 명확할 때’ 그리고, 일반적인 용도라면 유틸을 사용하라고 해서 사용자에게 입력을 받고, 해당 입력값에 대해서 도메인으로 바꾸고 하는 로직 등이 들어가다 보니까 util로 잡았던 것 같아요.

말씀해 주신 부분처럼, Input에서 Cars까지 생성하면서 도메인 객체의 생성까지 담당하게 되니까 그 역할이 조금 모호해졌다는 생각이 들었어요…! 리팩토링 하면서 이름까지만 입력받고, String 값을 리턴해서 **해당 Cars를 생성하는 부분은 컨트롤러에서 진행**하도록 하는 게 맞는 것 같아요! 좋은 지적 감사드립니다 🥰 (마찬가지로 Race 역시 그렇게 진행하는 게 맞을 것 같아요)

추가적으로 고민했던 부분에 대해서도 한 가지 질문드리고 싶어요! 기본적으로 컨트롤러에서 InputView 대신에 InputUtil을 참조하도록 코드를 작성하고 싶었는데요. 

`inputUtil.getUserInput(inputUtil::getCarNames)`

그러다 보니깐 이런 식으로 코드 상에서 InputUtil의 재입력 메서드의 인자인 Supplier에 들어갈 자동차 이름을 받는 부분을 별도의 메서드로 분리하면서 inputUtil을 2번 작성하도록 되더라구요…! 

1. 이런 식으로 코드를 작성하는 게 괜찮을까요?
2. 아니면 getCarName에 대한 부분은 별도로 view에서 진행하도록 할까요? 단, 이러면 컨트롤러에서 view와 util 모두 참조하면서 util이 진입점으로서 역할을 못할 것 같아서요…!
3. 혹은, 아예 재입력을 위한 클래스를 분리하고 inputUtil 대신에 그냥 view에서 처리하도록 진행하는 것이 나을까요? (진입점 개념을 제거)

페어와 진행했을 때는 1번이 더 나을 것 같다고 생각되어 1번으로 진행했는데, 또링의 의견도 들어보고 싶어서 이렇게 질문드려요! 🥹

jnsorn replied:
자세한 답변 고마워요! 😎
져니는 왜 InputView의 진입점(=InputUtil)이 필요하다고 느끼셨나요?

Cl8D replied:
사실 리팩토링 이전 초기 설계에서는 InputView는 콘솔을 통해서 입력받는 메서드만 두고, OutputView에서는 메시지를 콘솔에 출력하는 기능만 두고 싶었어요. InputUtil이 없으면 사용자에게 메시지를 출력하는 OutputView를 InputView에서 함께 참조하게 되니까 View끼리의 결합도가 높아진다고 판단해서 생성해두었어요. 즉, 일종의 InputView와 OutputView를 함께 사용하는 util 클래스 정도로 초기에는 설계했어요. 
- 이름이 InputUtil인 것은, 사용자에게 값을 입력받는 행위 자체가 ‘**사용자에게 메시지를 <입력 메시지>를 출력하고, 메시지를 입력받는다**’를 책임진다고 생각해서 InputUtil이라고 이름 지었습니다!

다만, 또링의 질문을 보고 생각하다 보니까 현재의 OutputView는 값을 출력해야 하는 값의 가공까지 맡고 있기 때문에 제가 진행한 초기 설계와는 맞지 않는 생각도 들기는 하네요…! 

또한, **OutputView의 메시지 출력을 Controller에서 진행하게 되면** (현재도 Controller에서 OutputView를 참조하고 있으니까요. 생각해보면 중복된 코드였네요...) ~~InputView에서 굳이 참조하지 않아도 되니, InputUtil을 제거하고 InputView에서 처리하도록 진행해도 충분히 괜찮을 것 같다는 생각이 들었어요!~~ 

-> 이렇게 생각했었는데, 리팩토링을 하다 보니까 에러 발생 시 재입력을 할 때 에러 메시지를 출력해줘야 되네요 🥲 그래서 우선 **에러 메시지를 출력하는 부분은 static 메서드를 통해서 처리**하도록 진행했습니다. 

-> 메시지 출력은 인자로 message를 받아서 System.out.println으로 출력만 해주면 되니까, static으로 선언하더라도 문제가 없을 것이라고 판단했어요! 진행하면서 OutputView의 모든 메서드를 static으로 처리할까도 고민했었는데, 무분별한 static 메서드의 사용은 메모리에 부하를 줄 수 있으니까 메시지 출력부만 그렇게 진행했습니다 😊

글로 생각을 정리하다 보니까 이런 고민들이 조금 더 쉽게 해결된 것 같아요. 리팩토링 할 때 InputUtil을 제거하고 InputView에서 처리하도록 진행하겠습니다. 좋은 지적 감사드려요 👍

jnsorn commented on src/main/java/racingcar/controller/Controller.java:
```diff
@@ -0,0 +1,6 @@
+package racingcar.controller;
+
+public interface Controller {
+
+    void process();
+}
```
Controller를 인터페이스로 분리하셨네요! 
이렇게 구현하면서 어떤 장점을 느끼셨는지 궁금합니다. 😊

Cl8D replied:
현재는 RaceController 하나여서 Controller에 대한 인터페이스를 구현하여도 제대로 활용하지 못하고 있는 것 같아요…! 

프리코스 때 경험을 생각해보면, **컨트롤러의 진입점에 대해서 인터페이스의 메서드로 선언**해두고 해당 컨트롤러에 대해 호출할 때 모두 동일하게 호출할 수 있었어서 코드상으로 깔끔하다고 생각이 들었습니다! 😊 

추가적으로, 단순히 인터페이스의 장점에 대해서 생각해보았을 때 많이들 말씀해주시는 ‘**선언과 구현의 분리**’를 이룰 수 있다는 점도 큰 장점인 것 같습니다. 또한, 외부적으로 모든 컨트롤러에 대한 공통적인 기능을 호출해야 할 때 인터페이스인 ‘Controller’를 통해서 한 번에 매개변수로 받고, 내부적으로 세부 로직을 호출할 수 있다는 것도 하나의 장점이 될 것 같습니다!

jnsorn replied:
인터페이스의 목적을 잘 알고계신것 같아요. 👍 
하지만 져니가 말씀하신대로 지금 코드에서는 인터페이스의 장점이 나타나지 않고 있어요. 
필요하지 않은 과한 설계는 코드의 복잡함을 높이기도 하니, 이 부분도 고민하며 코드를 작성하는 습관을 가지면 좋을 것 같아요.
그래도 인터페이스에 대해 학습하시고 적용해보신 것은 매우 좋습니다. 💯

Cl8D replied:
과한 설계는 코드의 복잡도를 증가시킨다는 점을 배울 수 있는 것 같아요. 이번 리팩토링 때 해당 부분은 수정하겠습니다! 감사해요 🥰

jnsorn commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,69 @@
+
+    public void move() {
+        drivingMark.append("-");
+    }
```
출력 요구사항이 아래와 같이 변경된다면 어떻게 될까요? 
```
pobi : ==
crong : =
honux : ==
```
View의 변경이 도메인에 영향을 끼칠 것 같은데 어떻게 수정해보면 좋을지 고민해봐요!

jnsorn commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public String toString() {
+        return name + " : " + drivingMark;
+    }
```
toString의 목적은 무엇이며, 위와 같이 재정의 했을 때 어떤 장단점이 있을지 고민해보면 좋겠네요. 😎

Cl8D replied:
toString에 대해서 조금 더 알아보고 싶어서, 이팩티브 자바 책이랑 [공식문서](https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#toString())를 확인해보았어요!

기본적으로 toString은 ‘간결하면서 사람이 읽기 쉬운 형태의 유익한 정보’를 반환해야 합니다. 본 책에서는 모든 하위 클래스에서 toString을 재정의하여 사용하라고 하고 있어요! 

`System.out.println(Car) // pobi:----`

저는 ‘해당 자동차에 대한 정보 출력 = 곧, 이름과 주행 거리에 대한 출력’이라고 생각하여 toString을 통해서 재정의하였고, 덕분에 출력할 때 ‘Car’ 자체만 사용하더라도 바로 출력을 할 수 있었어서 이런 부분은 편리했다고 느꼈던 것 같아요!

다만, toString의 기본 용도는 ‘해당 객체가 가지고 있는 모든 정보의 반환’을 위해서 사용해야 하기 때문에, 현재 코드에서는 각 필드의 값만 출력하고 있지, 해당 필드가 어떤 건지는 출력하지 않아 toString의 목적 자체에는 부합하지 않다고 생각됩니다. **자동차의 위치를 출력하는 로직은 뷰 쪽으로 이관**하여 해당 작업을 진행하는 것이 도메인 관점으로 봤을 때 더 깔끔할 것 같아 그렇게 리팩토링하려고 합니다!

jnsorn commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,94 @@
+            result.append(car).append("\n");
+        });
+        return result.toString();
+    }
```
Cars가 경주 결과를 그려주고 있네요! Cars의 역할이 맞을까요?

Cl8D replied:
코드를 확인하다 보니, race와 내부적으로 호출하는 checkNumberAndMove 역시 하나의 메서드에서 너무 많은 것을 담당하고 있다는 생각이 들었어요! StringBuilder를 통해 경주에 대한 결과를 넣어주는 부분은 도메인에서 처리하지 않고 **뷰에서 처리**하는 게 맞다고 생각이 들어서 이 부분은 옮기려고 합니다! 😎

jnsorn commented on src/main/java/racingcar/domain/Race.java:
```diff
@@ -0,0 +1,44 @@
+        int count;
+        try {
+            count = Integer.parseInt(tryCount);
+        } catch (NumberFormatException e) {
```
적절한 예외 catch! 👏🏻

jnsorn commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,39 @@
+        assertThatThrownBy(() -> Car.of(carName))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining(BLANK_MESSAGE.getValue());
+    }
```
instance와 message 비교로 더 구체적인 테스트가 되었네요. 👍

jnsorn commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,125 @@
+        // then
+        assertThat(car)
+                .isEqualTo(TestProvider.createTestMoveCar(car));
+    }
```
테스트 열심히 작성해주셨네요! 특히 경계값 테스트 아주 좋습니다. 😄

다만, 테스트가 지나치게 복잡하게 느껴지는데요 어떻게 개선할 수 있을까요?
자동차가 특정 조건에 의해 움직이는지 테스트를 하기위해 cars를 생성할 필요가 있을까요?
해당 테스트의 행위는 어떤 객체의 책임인지 고민해보면 좋을 것 같아요.

Cl8D replied:
사실 테스트 코드를 작성하면서도 private 메서드를 이런 식으로 테스트하는 것이 맞을까… 하고 정말 고민이 많았습니다 ㅠㅠ private 메서드는 public을 통해 테스트 해야 하는데, 메서드 내부에서 계속 연관지어 호출하다 보니까 단위 테스트라는 이름이랑 맞지 않게 구현이 된 것 같아요.

우선 리팩토링 시 **private 메서드는 테스트하지 않는 방향**으로 진행할 예정이고, 특히 리뷰 달아주신 테스트의 경우 자동차의 전진에 대한 내용이기 때문에 **‘CarTest’로 가는 게 더 부합**된다고 판단되어요! 아마 도메인 쪽 로직을 변경하다 보면 테스트 쪽 로직도 많이 변경될 것 같아서 함께 수정해보려고 해요!

- **테스트 하기 어려운 부분은 외부에서 받아올 수 있도록 진행**하기 → 랜덤 넘버를 생성해주는 부분에 대해서 Mock 클래스를 생성해서 setter를 통해 해당 랜덤 넘버를 고정시켜주는 방향으로 진행해볼까 싶어요!

- **도메인에 대한 역할을 명확**하게 하기 → 지금은 일급 컬렉션인 Cars에서 너무 많은 것들을 담당하고 있다는 생각이 들었어요…! 뷰와 도메인을 분리하고, 최대한 도메인이 해야 되는 일만 진행할 수 있도록 간결하게 진행해볼까 해요 😄

- 리뷰해 주셔서 감사합니다! **열심히 수정해서 더 보기 좋은 코드**로 PR 드릴게요… 😎

jnsorn replied:
테스트하기 좋은 코드를 위해 많은 고민을 하신 것 같아요 👍 말씀하신 방향 모두 좋습니다.

jnsorn reviewed (changes requested):
안녕하세요. 져니 👋 첫 미션인데 잘 구현해주셨어요!
특히 문서가 아주 잘 되어있어 빠르게 요구사항을 파악할 수 있었습니다.
전반적으로 메서드 분리 및 컨벤션은 잘 지켜주셔서, 져니의 의견이 듣고 싶은 부분에 코멘트를 남겼어요.
확인해보시고 궁금한 부분은 편하게 DM 주세요.

Cl8D committed:
refactor: 불필요한 파일 제거

Cl8D committed:
feat: 각 도메인별로 필요한 상수 클래스 분리

- 기존의 DomainConstant에서 도메인별로 필요한 상수들을 분리했습니다.

Cl8D committed:
refactor: 매직넘버 처리 및 제너레이터를 도메인으로 패키지 이동

Cl8D committed:
feat: 최상위 커스텀 예외 클래스 생성

- try-catch에서 잡을 때 모든 예외를 나열하는 것보다는 한 번에 정의하는 게 좋을 것 같아서 부모 예외 클래스를 생성하였습니다.

Cl8D committed:
feat: Car의 필드에 대한 wrapper 클래스 생성 및 검증 로직 이동 

- 원시값을 포장하기 위해서 별도의 wrapper 클래스를 생성하고, 추후 Car의 필드가 추가된다면 validation 로직이 car에 들어가있을 때 생성자가 매우 길어질 것이라고 판단해서 wrapper 클래스에서 검증하도록 이동했습니다.

Cl8D committed:
feat: Cars의 메서드 분리 및 출력과 관련된 부분 제거, CarRaceDto 사용

- 자동차 각각의 경주 결과에 대해서 담는 DTO를 생성하였으며, 기존에 StringBuilder를 통해 경주 결과를 리턴했던 로직을 뷰로 이관하였습니다.

Cl8D committed:
feat: EnumUtil의 제네릭 타입 추가 및 메시지 추가

- 기존에 Object로 받았던 것을 변경하여 제네릭 타입을 하나 더 선언하였습니다.

Cl8D committed:
feat: Race의 필드에 대한 wrapper 클래스 생성 및 경주 결과를 저장하는 Dto 추가

Cl8D committed:
fix: 재입력 시 catch하는 예외 수정 및 InputUtil에서 도메인을 생성하지 않도록 수정

Cl8D committed:
fix: 화면 출력에 대한 부분을 outputView에서 처리하도록 수정

Cl8D committed:
fix: 초기값이 제대로 안 나오는 이슈 수정 및 파라미터 전달 수정

- 뷰에서 자동차 경주의 초기값이나 경주를 시작하도록 하지 않고, 컨트롤러에서 시작해서 뷰에서는 단순히 결과만 출력하도록 로직 수정

Cl8D committed:
fix: 자동차의 전진 조건 판단을 cars -> car에서 하도록 수정

- 자동차가 전진하는 것의 책임은 'car' 도메인이 가지고 있다고 생각하여 수정하였습니다.

Cl8D committed:
refactor: 자동차 경주 결과를 Car에서 생성하도록 변경 및 필드명 변경

- tryCount -> raceCount로 이름을 변경하였습니다.
- equals에서도 매개변수에 대한 네이밍을 변경하였습니다.

Cl8D committed:
refactor: 메서드 및 필드명 수정

Cl8D committed:
refactor: 메서드 순서 변경

- public 사이에 private가 있더라도 연관된 로직이 붙어있도록 변경
- getter를 가장 하위로 변경

Cl8D committed:
test: 사용자가 입력한 시도 횟수만큼 경주 결과를 생성하는지 체크하는 테스트 추가

Cl8D committed:
test: 자동차의 전진에 대한 테스트를 CarsTest -> CarTest로 이동

Cl8D committed:
test: 경주 진행 결과 개수 및 초기 자동차 위치 테스트 추가, private method에 대한 테스트 제거

Cl8D committed:
test: 테스트에서 사용하기 위한 랜덤 번호 제너레이터 추가

- boolean이 false일 경우 계속 0을 리턴.
- 그게 아니라면 호출할 때마다 증가하는 값(1, 3, 7...)을 리턴하도록 추가

Cl8D committed:
docs: 도메인 설계서 수정

Cl8D committed:
fix: 에러 발생 시 재입력받을 수 있도록 수정 및 메서드 분리

Cl8D committed:
fix: 출력 형식에 대해 변경 진행

Cl8D committed:
feat: 자동차 이름에 대한 공백 제거 로직 추가 

- 중복값 검증 시 'pobi,pobi'는 가능하지만 'pobi, pobi'는 안 되어서 양끝의 공백을 제거하도록 진행하였습니다.

Cl8D committed:
test: 중복값 테스트 시 이름에 공백값이 들어가도록 걸러낼 수 있도록 이름 조건 추가

Cl8D commented:
안녕하세요! 두 번째 제출 전에 어디다가 커멘트를 남겨야 될지 모르겠어서 우선 여기에다가 달았습니다.
전반적으로 구조를 많이 바꾸다 보니까 커밋의 개수가 상당해서 리뷰 요청드리기가 민망하네요 🥹
따끔한 리뷰 부탁드려요 😋
리팩토링을 진행하면서 다음과 같은 것들을 최대한 고쳐보고자 했어요!

> 1. private method의 테스트는 최대한 **public을 통해 진행**하도록 만들기
> 2. 각 **객체에 대한 역할을 확실하게** 하기 (ex. 뷰에서 Cars, Race 같은 도메인을 생성하지 않도록 만들기)
> 3. **Wrapper 클래스** 활용하기 (각 필드에 대한 검증도 wrapper에서 진행하기)
> 4. **화면 출력은 view**에서 진행하도록 진행하기 
> 5. 메서드의 **순서 변경**하기 

리팩토링을 진행하면서 wrapper 클래스를 적극적으로 활용하려고 했는데, 이유는 생성자에서 validation 로직을 실행할 때 필드가 많아지면 많아질수록 생성자의 길이가 점점 더 길어지는 문제를 해결하기 위해서였어요.
다만, 걱정인 건 필드가 많아질수록 **각 도메인의 모든 필드마다 wrapper 클래스를 만들면 너무 낭비가 되지 않을까**...? 라는 생각이 들었어요.
혹시 이 부분에 대해서는 어떻게 생각하시는지 또링의 의견이 궁금해요! 😊

그리고, getter 사용을 지양해야 된다고 생각하지만 Car에서 getName을 할 때 wrapper 클래스인 CarName의 getName()을 하는 것이기 때문에 해당 필드를 한 단계를 거쳐서(?) 가게 되는 것이니까 조금 더 getter 사용이 완화되었다고 생각하는데, 괜찮은 설계인지도 궁금해요! 

마지막으로, 메서드의 순서에 대한 것인데요! 기존에는 메서드 위치를 public, private로 나누어서 public을 상단에, private를 아래로 내리는 구조를 채택하였는데, [검색](https://dodop-blog.tistory.com/277)을 해보니 public - private보다는 연관된 로직끼리 묶는 게 더 좋다는 글을 봤어요. public 사이에 private 메서드가 들어가기는 하는데, 또링은 어떤 구조를 선호하시는지 궁금해요!

리뷰해 주셔서 감사드립니다 🙇‍♀️ 잘 부탁드려요!
++ 하단에 몇 개의 커밋을 더 올렸는데... 코드에 중복이 많은 것 같아서 그것만 수정해서 추가 커밋 진행했습니다 :D

Cl8D committed:
refactor: 메서드명 변경, 변수 추출, 메서드 위치 변경 등 리팩토링 진행

Cl8D committed:
refactor: 테스트에 대한 중복 로직 리팩토링

Cl8D committed:
refactor: 테스트 매직 넘버 처리 추가

Cl8D committed:
refactor: 클래스명, 파라미터 네이밍 수정

Cl8D committed:
refactor: 메서드명 네이밍 수정

jnsorn commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,60 @@
+
+public class Car {
+
+    private final CarName name;
+
+    private final CarPosition position;
```
값을 유의미하게 감싸 사용하고 있네요 👍 
더불어 Car내에 흩어져있던 검증도 해당 객체내에서 진행함으로써, 객체의 책임이 명확하게 분리된 것 같아요.

jnsorn replied:
> 리팩토링을 진행하면서 wrapper 클래스를 적극적으로 활용하려고 했는데, 이유는 생성자에서 validation 로직을 실행할 때 필드가 많아지면 많아질수록 생성자의 길이가 점점 더 길어지는 문제를 해결하기 위해서였어요.
> 다만, 걱정인 건 필드가 많아질수록 각 도메인의 모든 필드마다 wrapper 클래스를 만들면 너무 낭비가 되지 않을까...? 라는 생각이 들었어요.
> 혹시 이 부분에 대해서는 어떻게 생각하시는지 또링의 의견이 궁금해요! 😊

좋은 고민이에요! 한 객체를 생성할 때 수행해야할 로직이 많아진다는 건, 해당 객체에 많은 책임이 부여되어있다고도 생각할 수 있어요. 그 관점에서 CarName과 Position을 추출하여 각 도메인의 책임을 분리한 것은 의미있는 분리라고 생각합니다. 👍 
하지만 져니가 말씀하신대로 도메인의 모든 필드마다 값을 감싼 객체를 만드는 건 비효율적이라고 생각이 들어요. 특히, 별도의 비즈니스 로직없이 그저 값을 감싸기만 한 객체는 오히려 복잡함만 증가시킨다고 생각합니다. 
따라서, 해당 필드를 감싸 사용했을때 어떤 장,단점이 있을지 생각해보면서 앞으로 져니만의 기준을 만들어 나가면 좋을 것 같아요 :)

Cl8D replied:
단순히 값을 감싸는 용도가 아닌, 검**증 로직이나 별도의 로직이 들어가 원본 객체의 책임을 덜 수 있다면** 분리하는 것이 더 좋을 것 같기는 해요. 혹은, 해당 객체가 **도메인에서 중요한 역할을 하는 필드**라면 감싸서 반환하는 것이 가독성이 증가할 것 같다는 생각이 들었어요 😋

또한, 현재 미션과는 관련이 없지만 이전에 사이드 프로젝트를 진행하면서 primitive 타입을 객체로 포장하여 진행하는 팀원을 본 적 있었어요. 그 팀원에게 물어보니까 값을 전달하거나 가공할 때 숫자값을 넣어서 전달하면, 비즈니스 로직이 복잡해질 때 개발을 진행하면서 숫자값을 가지는 다른 필드와 헷갈릴 수가 있어서 포장해서 사용하는 경우가 많다고 하더라고요. **코드상으로도 해당 값이 어떤 것을 의미하는지** 더 잘 알 수 있기도 하고요. 그때는 그게 비효율적이라는 생각이 들었었는데, 지금 생각해 보니까 어떤 의미로 말씀하신 건지 와닿는 것 같아요! 앞으로도 종종 써먹어야겠어요 😎

jnsorn commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,60 @@
+
+    private Car(final String name) {
+        this.name = CarName.of(name);
+        this.position = CarPosition.of();
```
정적 팩터리 메서드를 생성하셨네요! 👏🏻
네이밍을 많이 고민하셨을 것 같은데, [이 글](https://sun-22.tistory.com/84)을 참고해서 다시 네이밍해봐도 좋겠네요!

Cl8D replied:
저는 정적 팩터리 메서드 생성자를 사용할 때, 네이밍을 이 [레퍼런스](https://docs.oracle.com/javase/tutorial/datetime/overview/naming.html)를 참고해서 사용하고 있었어요. 오라클의 네이밍 컨벤션인데, 내부적으로는 Java Date-Time API에서 사용되는 내용이지만 코드에서도 충분히 사용할 수 있는 점이라고 생각되어서 지금까지 사용해 왔었어요. **파라미터로 넘어온 값들을 검증하여 인스턴스를 생성할 때 of를 흔히 사용**한다고 하더라구요!

근데 또링이 보내주신 레퍼런스도 읽어봤는데, 확실히 네이밍을 통해 명시적인 선언을 진행해주면 더 이해하기가 쉬울 것 같다는 생각이 드네요. (사실 개발할 때 네이밍이 중요한데, of로 통일하면 딱히 생각할 필요가 없으니까 더 그렇게 사용했던 것 같아요… 반성… 🥲) 또, 레퍼런스에서는 of는 집계 메서드에서 주로 사용한다는 것을 발견해버렸어요! 이번 리팩토링 때는 **create를 통해서 인스턴스 생성을 나타내도록 유의미한 네이밍**을 가지도록 수정해보겠습니다!

jnsorn commented on src/main/java/racingcar/controller/CarRaceController.java:
```diff
@@ -0,0 +1,47 @@
+        List<RaceResultDto> totalRaceResult = race.getRaceResults(cars);
+        outputView.printRaceResult(initCarResult, totalRaceResult);
+
+        List<String> winners = cars.pickWinners();
```
controller에서 바로 cars를 이용해 우승자를 뽑고 있어요. 
경기를 관리하는 클래스가 있으니, 거기에 우승자를 선출하는 역할을 부여하면 어떨까요?

Cl8D replied:
리팩토링을 진행하면서 이 부분에 대해서 의문점이 들어서 커멘트 남겨요!
현재는 컨트롤러에서 바로 도메인을 참조하지 않고, 중간 계층인 CarsService를 두어서 해당 서비스에서 우승 결과에 대해서 바로 리턴하도록 리팩토링을 진행했어요.

` public List<String> pickWinners(final Cars carsInfo) {
        return carsInfo.pickWinners();
    }
`

위 코드는 컨트롤러에서 도메인을 바로 접근하도록 하지 않기 위함이기 때문에 (혹은, 추후 비즈니스 로직이 추가된다면 뽑은 winner 리스트를 가공한다든지, 그러한 로직이 추가되었을 때 해당 서비스 단에서 처리하기 위해서) 현재는 코드가 cars를 바로 호출하는 식으로 진행되었어요. 

다만, 말씀해 주신 경기를 관리하는 클래스는 Race를 말씀하신 것 같은데, 왜 Race에서 우승자를 산출하는 역할을 부여해야 하는지 잘 이해가 되지 않아서 질문드려요! 일급 컬랙션인 cars에서 자동차 리스트에 대한 정보를 가지고 있기 때문에, **외부에서 getter를 호출하지 않고** 해당 객체를 통해 바로 maxPosition을 뽑은 다음 (maxPosition() 메서드에 대해 private로도 유지할 수 있게 되어요.) filter를 통해서 우승자 리스트를 뽑을 수 있는데, 만약 Race에서 관리하게 된다면 getCars()를 통해 car의 정보를 받아오고, 또 cars에게 maxPosition 정보를 받아온 다음 정보를 처리하는 식으로 진행해야 될 것 같아서요.

이렇게 되면 '**객체에게 질문을 던지는 것**'이 아니라, 외부에서 getter를 통해 받아오고, 그리고 다시 가공하는 식으로 로직이 짜여진다고 생각이 되는데, 제가 혹시 또링의 의견을 잘못 이해한 걸까요? 🥹...

Race에서는 단순히 사용자가 입력한 시도 횟수를 관리하고 있기 때문에, 추가적으로 우승을 위한 maxPosition을 관리하는 필드를 두어서 구현할까도 고민해보았지만, maxPosition을 계속 업데이트하는 로직이 들어가야 하고, maxPosition의 로직 자체가 **cars 리스트의 위치 중에서 가장 큰 위치**를 의미하고 있기 때문에 Race에 두는 건 별로 좋지 않다고 생각했어요!

또링은 이에 대해서 어떻게 생각하시는지 궁금해요!

jnsorn commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,90 @@
+        return carNames.split(SPLIT_DELIMITER.getValue());
+    }
+
+    private void validateDuplicateCarName() {
+        int nonDuplicateCount = new HashSet<>(cars).size();
+        if (cars.size() != nonDuplicateCount) {
+            throw new DuplicateException(DUPLICATE_MESSAGE.getValue());
+        }
+    }
```
자료구조의 특성을 이용한 중복 확인! 👍 
(참고) 개인적으로 부정형의 네이밍은 가독성이 떨어진다 생각해서 저라면 uniqueCount와 같은 단어를 사용했을 것 같아요!

jnsorn commented on src/main/java/racingcar/controller/CarRaceController.java:
```diff
@@ -0,0 +1,47 @@
+        Cars cars = getUserInputAndCreateCars();
+        Race race = getUserInputAndCreateRace();
+
+        List<CarRaceDto> initCarResult = cars.initStatus();
```
`initStatus()`는 자동차들을 초기 상태로 만들어주는 메서드라고 생각되는데, DTO를 반환하고 있네요. 😮
도메인 내부에서 DTO를 생성하면 어떤 장단점이 있을까요?

Cl8D replied:
슬랙의 질의응답에서도 나왔던 주제였어서 또링에게 한 번쯤 피드백 받을 내용이라고 생각했는데, 이렇게 딱 질문을 주시네요…! 우선 initStatus는 자동차를 초기 상태로 만들어주는 것이 아닌, **자동차의 초기 상태를 가져오기 위한** 메서드였어요. 네이밍이 init으로 시작하다 보니까 초기화라고 생각하신 것 같아요! 이 부분은 수정하겠습니다. 🥹

도메인 내부에서 Dto를 생성하게 된다면, **Dto의 생성자가 변하게 되었을 때 도메인 코드 역시 함께 변경**된다는 단점이 존재해요. Dto의 경우 도메인에 비해 변경이 잦은 부분이기 때문에, 도메인 로직 역시 함께 변경된다는 점이 좋지 않다고 생각이 들어요. 또한, DTO는 이름 그대로 Data Transfer Object로서, 계층 간의 데이터를 전달해주는 역할을 하는 객체인데 **도메인이 DTO를 의존하게 되는 건 이상하다**고 생각이 들어요.

장점이라고 한다면, 외부에서 getter를 통해 값을 접근하지 않고 내부의 필드값을 통해서 바로 DTO를 생성할 수 있다는 장점이 있을 것 같아요. (getter를 지양한다는 관점 하에서...! 이 부분은 생각해봐야 될 문제 같아요. 지금까지 너무 getter를 무작정으로 안 쓰려고 했던 것 같아서요. ㅎㅎ)

하지만 전자의 경우가 리팩토링이 더 필요한 부분이라고 생각이 들어요. **별도의 서비스 계층을 만들어서 DTO를 반환**하도록 하고, 이 결과를 컨트롤러에서 사용하도록 하는 게 더 좋은 구조가 아닐까… 라는 생각이 들었어요. 리팩토링 하면서 한 번 수정해보겠습니다! 😊

jnsorn commented on src/main/java/racingcar/controller/CarRaceController.java:
```diff
@@ -0,0 +1,47 @@
+        Race race = getUserInputAndCreateRace();
+
+        List<CarRaceDto> initCarResult = cars.initStatus();
+        List<RaceResultDto> totalRaceResult = race.getRaceResults(cars);
```
하나의 메서드가 두 가지 일을 하고 있는 것 같아요!
하나의 메서드는 하나의 행위만 하도록 수정해볼까요?

jnsorn commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,60 @@
+        return Objects.hash(name);
+    }
+
+    public String getName() {
+        return name.getName();
+    }
```
> 그리고, getter 사용을 지양해야 된다고 생각하지만 Car에서 getName을 할 때 wrapper 클래스인 CarName의 getName()을 하는 것이기 때문에 해당 필드를 한 단계를 거쳐서(?) 가게 되는 것이니까 조금 더 getter 사용이 완화되었다고 생각하는데, 괜찮은 설계인지도 궁금해요!

답변드리기에 앞서, 져니는 왜 getter를 지양해야 한다고 생각하세요?

Cl8D replied:
객체를 더 객체스럽게 사용하기 위해서는, 객체에게 값을 받아오는 것이 아닌 **객체에게 직접 질문을 던지는 형태**로 이루어져야 된다고 생각하기 때문이에요. **객체가 가지고 있는 값을 객체 스스로 판단**하고, 외부에 대해서 최대한 영향을 덜 받도록 하고 싶어서 getter / setter를 지양하려고 하는 편이에요. 

하지만, 생각해 보니까 무작정으로 getter를 지양하는 것도 별로 좋은 설계라는 생각이 들지 않았어요. 특히 이전처럼 getter를 사용하지 않기 위해서 도메인 단에서 DTO를 리턴하는 등의 행위는 오히려 **객체 간의 결합도를 증가**시키는 행위라고 생각이 들었어요...! 그래서 저만의 작은 기준을 세웠는데, **출력을 위해 필요하거나 DTO 변환** 등에서는 getter의 사용이 필수적이라고 생각이 되기 때문에 이런 경우에는 사용하는 게 더 낫다고 생각이 들어요!

그래서 이번 리팩토링 때는 cars에 대한 getter를 사용하면서 DTO 생성을 도메인에서 처리하지 않도록 변경하였어요. 또한, unmodifableList를 통해서 외부에서 변경하지 않도록 진행했습니다! 😊 이런 식으로 getter를 사용하는 건 괜찮을까요?

jnsorn commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,90 @@
+package racingcar.domain;
```
> 마지막으로, 메서드의 순서에 대한 것인데요! 기존에는 메서드 위치를 public, private로 나누어서 public을 상단에, private를 아래로 내리는 구조를 채택하였는데, [검색](https://dodop-blog.tistory.com/277)을 해보니 public - private보다는 연관된 로직끼리 묶는 게 더 좋다는 글을 봤어요. public 사이에 private 메서드가 들어가기는 하는데, 또링은 어떤 구조를 선호하시는지 궁금해요!

저는 public 메서드들 다음에 private 메서드들을 선언하는 것을 선호해요. 연관된 메서드끼리 묶으면 져니가 말씀하신대로 가독성이 높아진다는 장점이 있지만, 결국 이 객체가 외부에 제공하는 행위는 무엇인가(=public 메서드)를 파악하기에는 어려워진다는 단점이 있어서요. 또 요즘엔 IDE에서 public 메서드 내의 private 메서드들에 단축키를 통해 쉽게 접근할 수 있기 때문에 그렇기도 하구요. 이건 제 생각이니 져니도 져니만의 기준을 세워보세요! ㅎㅎ

Cl8D replied:
생각해 보니 Intelli만 하더라도 단축키로 접근할 수 있어서 public - private로 나누는 게 더 좋을 것 같네요. 특히 또링이 말씀해주신 것처럼 객체가 외부에 제공하는 행위에 대해서 파악할 수 있다는 게 더 좋은 것 같아요! 이 관점에 대해서는 생각 못하고 단순하게 다른 사람들이 public - private 순서로 쓰니까 저도 써야겠다고 생각했던 건데, 좋은 시야를 주셔서 감사드려요!! 👍

jnsorn reviewed (approved):
안녕하세요, 져니! 리뷰 잘 반영해주셨네요. 
많은 고민을 하면서 리팩터링 해주신 것 같아요. 특히 테스트가 더 깔끔하고 풍부해졌어요. 👍 
추가로 더 이야기해보면 좋을 부분에 대해 코멘트 남겨두었는데, 이건 2단계 진행하면서 같이 답변달아주세요!
이번 단계는 여기서 마무리하겠습니다 😄
